float32[] phases
float32[] offsets
float32[] amplitudes